src/values/ray-query.cxx

```diff
@@ -0,0 +1,290 @@
+
+public:
+    RayQuery(Type t): Value(t), tlas(t), active(false) {}
+
+    /// @brief Must be called before tracing a ray in a ray query.
+    /// Initializes the ray query with an acceleration structure and ray information.
+    /// @param as acceleration structure to query.
+    /// @param ray_flags ray flags.
+    /// @param cull_mask ray cull mask.
+    /// @param origin world-space ray origin.
+    /// @param direction world-space ray direction.
+    /// @param t_min ray interval lower-bound.
+    /// @param t_max ray interval upper-bound.
+    void initialize(
+        const AccelerationStructureManager& as,
+        const unsigned ray_flags,
+        const unsigned cull_mask,
+        const std::vector<float>& origin,
+        const std::vector<float>& direction,
+        const float t_min,
+        const float t_max
+    ) {
+
+        assert(origin.size() == 3);
+        assert(direction.size() == 3);
+
+        tlas = as;
+        this->rayFlags = ray_flags;
+        this->cullMask = cull_mask;
+        rayOrigin = glm::vec4(origin[0], origin[1], origin[2], 1.0f);
+        rayDirection = glm::vec4(direction[0], direction[1], direction[2], 0.0f);
+        rayTMin = t_min;
+        rayTMax = t_max;
+
+        active = true;
+        tlas.initStepTraceRay(rayFlags, cullMask, origin, direction, rayTMin, rayTMax, false);
+    }
+
+    /// @brief Take one step in tracing the ray where each step reaches the next geometry.
+    /// @return true if there is more to trace, otherwise false.
+    bool proceed() {
+        if (active) {
+            active = tlas.stepTraceRay();
+        }
+        return active;
+    }
+
+    /// @brief Make this ray query inactive.
+    void terminate() {
+        active = false;
+    }
+
+    /// @brief Generate and commit an intersection at <hit_t>. Only works if the candidate
+    /// intersection type is AABB.
+    /// @param hit_t distance from ray to intersection to consider in determining the closest hit.
+    void generateIntersection(float hit_t) {
+        tlas.generateIntersection(hit_t);
+    }
+
+    /// @brief Commit current candidate triangle intersection to be considered in determining
+    /// the closest hit. Only works if the candidate intersection type is triangle.
+    void confirmIntersection() {
+        tlas.confirmIntersection();
+    }
+
+    /// @brief Get the ray's t-min; ray's minimum interval.
+    /// @return ray's t-min.
+    float getRayTMin() const {
+        return rayTMin;
+    }
+
+    /// @brief Get the ray's flags.
+    /// @return ray flags.
+    unsigned getRayFlags() const {
+        return rayFlags;
+    }
+
+    /// @brief Get the ray's world-space origin.
+    /// @return world-space ray origin.
+    std::vector<float> getWorldRayOrigin() const {
+        return std::vector<float> {rayOrigin.x, rayOrigin.y, rayOrigin.z};
+    }
+
+    /// @brief Get the ray's world-space origin.
+    /// @return world-space ray origin.
+    glm::vec3 getWorldRayOriginGLM() const {
```

Comment:
Please remove this unused function and getWorldRayDirectionGLM() unless you plan to use them in the next update.